src/dmd/dtemplate.d

```diff
@@ -1351,8 +1351,98 @@ extern (C++) final class TemplateDeclaration : ScopeDsymbol
             {
                 Parameter fparam = fparameters[parami];
 
+                /* https://issues.dlang.org/show_bug.cgi?id=16486
+                 * If 'prmtype' is template alias, resolve it,
+                 * to the instance it's aliased to before proceeding.
+                 */
+                Type prmtype = fparam.type;
+
+                // We might have cyclic template references, so 'a' aliases to 'b'
+                // and 'b' goes back to 'a'. For now, we save the previous type
+                // and test it against (if they're equal) the new type.
+                // TODO: In the general
+                // case, we probably need full DFS cycle detection.
```

Comment:
One common "trick/hack/easy solution" in other parts of DMD is to use an iteration counter with a cut-off (e.g. 1000).

Reply:
Or he can use the `inuse` field
https://github.com/dlang/dmd/blob/a938d74e180cb87dc9c28ddf1f87f06eefd1dd45/src/dmd/declaration.d#L293
https://github.com/dlang/dmd/blob/a938d74e180cb87dc9c28ddf1f87f06eefd1dd45/src/dmd/dtemplate.d#L535
https://github.com/dlang/dmd/blob/a938d74e180cb87dc9c28ddf1f87f06eefd1dd45/src/dmd/dtemplate.d#L5951

Reply:
Thanks, those can come in handy. It also seems that the problem does not require full cycle detection because there's always only one way to go, meaning it's like floyd's tortoise and hare. I don't know if this is going to be useful though..